src/lsm/block_free_set.zig

```diff
@@ -1,37 +1,642 @@
 const std = @import("std");
 const assert = std.debug.assert;
 const mem = std.mem;
+const DynamicBitSetUnmanaged = std.bit_set.DynamicBitSetUnmanaged;
+const RingBuffer = @import("../ring_buffer.zig").RingBuffer(
+    usize,
+    64 / @sizeOf(usize), // 1 cache line of recently-freed blocks
+    .array,
+);
 
 /// The 0 address is reserved for usage as a sentinel and will never be returned
 /// by acquire().
-/// TODO add encode/decode function for run-length encoding to/from SuperBlock.
+///
+/// Set bits indicate free blocks, unset bits are allocated.
 pub const BlockFreeSet = struct {
-    /// Bits set indicate free blocks
-    free: std.bit_set.DynamicBitSetUnmanaged,
+    // Each bit of `index` is the OR of `shard_size` bits of `blocks`.
+    // That is, if a shard has any free blocks, the corresponding index bit is set.
+    index: DynamicBitSetUnmanaged,
+    blocks: DynamicBitSetUnmanaged,
+    // Track recently freed blocks. Stores 0-index bits, not 1-indexed block addresses.
+    recent: RingBuffer = .{},
 
-    pub fn init(allocator: *mem.Allocator, count: usize) !BlockFreeSet {
+    // Fixing the shard size to a constant rather than varying the shard size (but
+    // guaranteeing the index always a multiple of 64B) means that the top-level index
+    // may have some unused bits. But the shards themselves are always a multiple of
+    // the word size. In practice the tail end of the index will be accessed less
+    // frequently than the head/middle anyway.
+    //
+    // 10TiB disk / 64KiB block size / 640B shard size = 4096B index
+    const shard_size = 10 * (64 * 8); // 10 cache lines per shard
+
+    pub fn init(allocator: *mem.Allocator, total_blocks: usize) !BlockFreeSet {
```

Comment:
Thanks for the feedback! Could I assert that the `total_blocks` is always a multiple of the `shard_size`, or is that too aggressive? It would simplify some things, e.g. `grow`.

Reply:
Nevermind, I just noticed another PR comment where you suggested the same thing :slightly_smiling_face: